### PR TITLE
Multi window close

### PR DESCRIPTION
**Pull request by Redth**
Adds:

- `IApplication.OnWindowClosed`

This now tracks when a UISceneSession disconnects, an Activity is destroyed, and a Window is closed and remove its Window from the Application's collection of windows.

- `IApplication.CloseWindow`

This method provides a way to request that an `IWindow`'s platform scene/activity/window is closed intentionally without the user requesting it specifically through the platform's mechanism to do so.

- [x] TODO: Windows support for detecting the window is closed
- [x] Question: Is the `OnWindowClosed` mapper the right way to do this?

### CODE REVIEW

**Comment by mattleibow** [comment]
Why is the window event on the app? Maybe it should be on the window? And then maybe exposed as an event as well?

**Comment by AmrAlSayed0** [comment]
I agree with @mattleibow, exposing window close as an event would allow multiple subscribers instead of having to subclass. The OnWindowClose approach feels like a workaround for languages that don't have events (eg. Java) but C# does.

Also, Android's On destroy is unreliable if you want to detect if the Activity is being closed. Android calls it if there is configuration changes and it might not even be called at all if the app/activity is killed due to memory constraints.

You can call `IsChangingConfigurations` or `IsFinishing` to see if the Activity is _actually_ being closed.

**Comment by Redth** [comment]
@AmrAlSayed0 thanks for the tip on that one, changed up where we call it :)

File: src/Controls/src/Core/HandlerImpl/Application.Impl.cs [hunk 1]
```diff
@@ -57,6 +57,31 @@ void IApplication.OpenWindow(IWindow window)
...
+			Handler?.Invoke(nameof(IApplication.CloseWindow), window);
+		}
+
+		internal void RemoveWindow(Window window)
+		{
+			// Window was closed, stop tracking it
+			if (window is null)
+				return;
+
+			if (window is NavigableElement ne)
```
Comment (Redth) [inline src/Controls/src/Core/HandlerImpl/Application.Impl.cs:71]: @PureWeen does this seem ok?

File: src/Core/src/Handlers/Application/ApplicationHandler.Standard.cs [hunk 2]
```diff
@@ -8,5 +8,7 @@ public partial class ApplicationHandler : ElementHandler<IApplication, object>
 
 		public static void MapTerminate(ApplicationHandler handler, IApplication application, object? args) { }
 		public static void MapOpenWindow(ApplicationHandler handler, IApplication application, object? args) { }
+		public static void MapCloseWindow(ApplicationHandler handler, IApplication application, object? args) { }
+		public static void MapOnWindowClosed(ApplicationHandler handler, IApplication application, object? args) { }
```
Comment (Redth) [inline src/Core/src/Handlers/Application/ApplicationHandler.Standard.cs:12]:
```suggestion
```

File: src/Core/src/Hosting/LifecycleEvents/AppHostBuilderExtensions.iOS.cs [hunk 3]
```diff
@@ -36,6 +37,13 @@ static void OnConfigureLifeCycle(IiOSLifecycleBuilder iOS)
 				.WillTerminate(app =>
 				{
 					app.GetWindow()?.Destroying();
```
Comment (Redth) [inline src/Core/src/Hosting/LifecycleEvents/AppHostBuilderExtensions.iOS.cs:39]:
```suggestion
					// By this point if we were a multi window app, the GetWindow would be null anyway
					app.GetWindow()?.Destroying();
```

File: src/Core/src/Platform/Android/MauiAppCompatActivity.cs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 using Android.OS;
 using AndroidX.AppCompat.App;
+using Microsoft.Maui.LifecycleEvents;
```
Comment (Redth) [inline src/Core/src/Platform/Android/MauiAppCompatActivity.cs:3]:
```suggestion
```

File: src/Core/src/Handlers/Application/ApplicationHandler.iOS.cs [hunk 5]
```diff
@@ -23,6 +23,24 @@ public static void MapOpenWindow(ApplicationHandler handler, IApplication applic
 			handler.NativeView?.RequestNewWindow(application, args as OpenWindowRequest);
 		}
 
+		public static void MapCloseWindow(ApplicationHandler handler, IApplication application, object? args)
+		{
+#if __MACCATALYST__ || __IOS__
+			if (args is IWindow window)
+			{
+				// See if the window's handler has an associated UIWindowScene and UISceneSession
+				var sceneSession = window.Handler?.MauiContext?.GetNativeWindow()?.WindowScene?.Session;
```
Comment (mattleibow) [inline src/Core/src/Handlers/Application/ApplicationHandler.iOS.cs:32]: window.Handler.NativeView should be UIWindow... Can you use that?

**Review by mattleibow — APPROVED** [review]
Feels good. Besides the nav area that @PureWeen should give an eyeball to just to be safe.

File: src/Controls/src/Core/HandlerImpl/Application.Impl.cs (hunk 1 above)
Comment (PureWeen) [reply src/Controls/src/Core/HandlerImpl/Application.Impl.cs:71]: Looks good 👍

**Pull request merged**